presto-pinot/presto-pinot-lib/src/main/java/com/facebook/presto/pinot/PinotConnection.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.pinot;
+
+import com.facebook.presto.pinot.PinotClusterInfoFetcher.TimeBoundary;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.apache.pinot.common.data.Schema;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.cache.CacheLoader.asyncReloading;
+import static java.util.Objects.requireNonNull;
+
+public class PinotConnection
+{
+    private static final Object ALL_TABLES_CACHE_KEY = new Object();
+
+    private final LoadingCache<String, List<PinotColumn>> pinotTableColumnCache;
+    private final LoadingCache<Object, List<String>> allTablesCache;
+    private final PinotConfig pinotConfig;
+    private final PinotClusterInfoFetcher pinotClusterInfoFetcher;
+
+    @Inject
+    public PinotConnection(PinotClusterInfoFetcher pinotClusterInfoFetcher, PinotConfig pinotConfig, @ForPinot Executor executor)
+    {
+        this.pinotConfig = pinotConfig;
+        final long cacheExpiryMs = this.pinotConfig.getMetadataCacheExpiry().roundTo(TimeUnit.MILLISECONDS);
+        this.pinotClusterInfoFetcher = requireNonNull(pinotClusterInfoFetcher, "cluster info fetcher is null");
+        this.allTablesCache = CacheBuilder.newBuilder()
```

Comment:
Probably just use:
```
private final Supplier<List<String>> tables = memoizeWithExpiration(this::getTables, metadataCacheExpiryMs, MILLISECONDS);
```

Reply:
Not quite. The key behavior I want here is asynchronous refresh. I don't want to pay the metadata access hit during a query. We did that before and that lead to these spiking queries every K minutes.